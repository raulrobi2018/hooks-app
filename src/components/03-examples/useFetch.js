import React, {useState, useEffect} from "react";

export const useFetch = (url) => {
    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null
    });

    useEffect(() => {
        // Para que cada vez que cargue un quote nuevo muestre el mensaje de loading,
        // vuelvo a setear el state
        setState({
            data: null,
            loading: true,
            error: null
        });

        fetch(url)
            .then((resp) => resp.json())
            .then((data) => {
                // Una vez que obtenemos la data, seteamos los valores
                setState({
                    data,
                    loading: false,
                    error: null
                });
            });
    }, [url]);

    return state;
};
