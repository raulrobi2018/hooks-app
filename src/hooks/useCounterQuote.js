import {useState} from "react";

// initialState con valor por defecto de 10
export const useCounterQuote = (initialState = 10) => {
    const [counter, setCounter] = useState(initialState);

    const increment = () => {
        setCounter(counter + 1);
    };

    const decrement = () => {
        counter > 1 ? setCounter(counter - 1) : setCounter(1);
    };

    const reset = () => {
        setCounter(initialState);
    };

    return {
        counter,
        increment,
        decrement,
        reset
    };
};
