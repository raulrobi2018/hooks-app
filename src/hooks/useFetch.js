import React, {useState, useEffect, useRef} from "react";

export const useFetch = (url) => {
    // La idea de este useRef es que mantenga la refencia mientras el componente
    // está montado. Cuando sea false, se cancelará las peticiones que se hayan hecho
    const isMounted = useRef(true);
    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null
    });

    // Cuando el componente se desmonte, se cambiará el valor de isMounted a false
    useEffect(() => {
        return () => {
            isMounted.current = false;
        };
    }, []);

    useEffect(() => {
        // Para que cada vez que cargue un quote nuevo muestre el mensaje de loading,
        // vuelvo a setear el state
        setState({
            data: null,
            loading: true,
            error: null
        });

        fetch(url)
            .then((resp) => resp.json())
            .then((data) => {
                // Solo seteará los datos si el componente está montado
                if (isMounted.current) {
                    // Una vez que obtenemos la data, seteamos los valores
                    setState({
                        data,
                        loading: false,
                        error: null
                    });
                }
            })
            .catch(() => {
                setState({
                    data: null,
                    loading: false,
                    error: "No se pudo cargar la info"
                });
            });
    }, [url]);

    return state;
};
