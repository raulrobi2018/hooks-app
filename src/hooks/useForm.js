import React, {useEffect, useState} from "react";

export const useForm = (initialState = {}) => {
    const [values, setvalues] = useState(initialState);

    // Desestructuro target del event
    const handleInputChange = ({target}) => {
        setvalues({
            ...values,
            [target.name]: target.value
        });
    };

    const {email} = values;

    useEffect(() => {
        console.log("CambiÃ³ el email");
    }, [email]);

    const handleSubmit = (e) => {
        e.preventDefault();
        console.log(values);
    };

    return [values, handleInputChange, handleSubmit];
};
