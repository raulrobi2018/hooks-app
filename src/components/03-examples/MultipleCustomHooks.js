import React from "react";
import {useFetch} from "./useFetch";
import "./styles.css";
import {useCounterQuote} from "../../hooks/useCounterQuote";

export const MultipleCustomHooks = () => {
    const {counter, increment, decrement, reset} = useCounterQuote(1);

    const {loading, data} = useFetch(
        `https://www.breakingbadapi.com/api/quotes/${counter}`
    );

    // !!data: conviente el null a un booleano y lo niega. !data lo convierte a booleano y luego !!data lo niega
    // Si es true, devuelve data[0]
    const {author, quote} = !!data && data[0];

    return (
        <div>
            <h4>BrakingBad Quotes</h4>
            <hr />

            {loading ? (
                <div className="alert alert-info text-center">Loading...</div>
            ) : (
                <blockquote className="blockquote">
                    <p className="mb-1">
                        {counter} - {quote}
                    </p>
                    <footer className="blockquote-footer mt-1">{author}</footer>
                </blockquote>
            )}

            <div className="row">
                <div className="col-4">
                    <button
                        className="btn btn-secondary"
                        onClick={() => decrement()}
                    >
                        Previous quote
                    </button>
                </div>
                <div className="col-4">
                    <button
                        className="btn btn-primary ml-1"
                        onClick={() => increment()}
                    >
                        Next quote
                    </button>
                </div>
                <div className="col-4">
                    <button
                        className="btn btn-success ml-1"
                        onClick={() => reset()}
                    >
                        Reset
                    </button>
                </div>
            </div>
        </div>
    );
};
