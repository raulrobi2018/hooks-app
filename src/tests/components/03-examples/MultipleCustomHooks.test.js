import React from "react";
import {MultipleCustomHooks} from "../../../components/03-examples/MultipleCustomHooks";
import {shallow} from "enzyme";
import {useFetch} from "../../../hooks/useFetch";
import {useCounterQuote} from "../../../hooks/useCounterQuote";

//Para que me de la ayuda de funciones hago esta importación
import "@testing-library/jest-dom";

//Lo que hace esto es que cuando se vaya a utilizar el useFetch en este archivo
//en vez de utilizar el useFetch, va a utilizar una implementación que vamos a inventar
jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounterQuote");

describe("Testing MultipleCustomHooks", () => {
    useCounterQuote.mockReturnValue({counter: 10});

    test("should display correctly", () => {
        //retorna lo que retorna el useFetch por defecto
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test("should display the data", () => {
        useFetch.mockReturnValue({
            data: [
                {
                    author: "Raul",
                    quote: "Hola mundo"
                }
            ],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        console.log(wrapper.html());

        expect(wrapper.find("alert").exists()).toBe(false);
        expect(wrapper.find(".mb-1").text().trim()).toBe("Hola mundo");
        expect(wrapper.find("footer").text().trim()).toBe("Raul");
    });
});
