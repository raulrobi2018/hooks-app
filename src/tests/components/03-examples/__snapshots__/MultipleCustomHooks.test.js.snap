// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MultipleCustomHooks should display correctly 1`] = `
<div>
  <h4>
    BrakingBad Quotes
  </h4>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <button
        className="btn btn-secondary"
        onClick={[Function]}
      >
        Previous quote
      </button>
    </div>
    <div
      className="col-4"
    >
      <button
        className="btn btn-primary ml-1"
        onClick={[Function]}
      >
        Next quote
      </button>
    </div>
    <div
      className="col-4"
    >
      <button
        className="btn btn-success ml-1"
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
